File: Cpp/543_Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int result = 0;
        getDepth(root, result);
        return result;
    }
    
    int getDepth(TreeNode* node, int& result){
        if(node == nullptr) return 0;
        int left = getDepth(node->left, result);
        int right = getDepth(node->right, result);
        result = max(result, left+right);
        
        return max(left, right) + 1;
    }
};

File: Kotlin/543_Diameter_of_Binary_Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        val (maxDepth, diameter) = recursive(root)
        
        return diameter
    }
    
    fun recursive(node: TreeNode?): Pair<Int, Int> {
        if(node==null) return Pair(0, 0)
        
        val (maxDepthLeft, diameterLeft) = recursive(node.left)
        val (maxDepthRight, diameterRight) = recursive(node.right)

        var maxDepth = maxOf(maxDepthLeft, maxDepthRight) + 1
        var diameter = maxOf(diameterLeft, diameterRight, maxDepthLeft + maxDepthRight)

        return Pair(maxDepth, diameter)
    }
}